### DISCUSSION TITLE
Setting up the java checkstyle linter

### DISCUSSION THREAD

**Discussion by 231tr0n** [nvimdev/guard.nvim#180] [Q&A] [answered]
### Feature description

1. ``require("guard.filetype")("java"):lint("checkstyle")`` - There is no way to run the linter if a formatter is not configured.
2. Provide ``Guard lint`` to only lint and not format file. Also make ``Guard fmt`` to only format and not lint file.
3. Can I have a combination where linter is run asynchronously and formatter is run synchronously.
4. Provide options for configuring async nature of linters and formatters in ``vim.g.guard_config``.

**Comment by xiaoshihou514** [comment]
>    There is no way to run the linter if a formatter is not configured.

You can [hook into the linter logic](https://github.com/nvimdev/guard.nvim/blob/main/ADVANCED.md#custom-logic-with-linters) in any way you want
 
>    Provide `Guard lint` to only lint and not format file.

Are there occasions where you _want_ to do this manually? Please describe them.

>    Can I have a combination where linter is run asynchronously and formatter is run synchronously.

Sync stuff is not normally what people want, what's the motivation?

>    Provide options for configuring async nature of linters and formatters in `vim.g.guard_config`.

Same as above

> Also a small doubt, I wanted know to why was the setup function deprecated. Most people generally expect for a setup function. Is there any chance to remove the vim.deprecate function in the setup?

You can read more about the motivation [here](https://github.com/nvim-neorocks/nvim-best-practices) :)

**Comment by 231tr0n** [comment]
@xiaoshihou514 Thank you for responding to my queries first of all.

1. ``require("guard.filetype")("java"):lint("checkstyle")`` - For the first point what I meant was that the configured linter doesnt run and throws a message saying no configured formatters. Basically, I just configured java checkstyle for linting with no formatter and ran ``Guard fmt`` and got a error message saying ``formatter not configured`` or something of that sort.
2. I wanted this feature because at my work, there are many cases where I should not touch the formatting of files since it changes git blame to my name even though I essentially made no change. But I still need the functionality to lint without formatting. That was my intention behind ``Guard fmt`` and ``Guard lint`` change.
3. For the third point, I wanted this feature because, let us just say a formatter takes time to format and other edits happen in the buffer, they get replaced by the text from formatter. Hence, I generally run it in sync mode. Linters can be run in async mode but formatters work best for me when run in sync mode.

**Comment by xiaoshihou514** [comment]
Basically, you want to only run checkstyle manually and infrequently, but guard doesn't provide configs for events.

**Comment by 231tr0n** [comment]
> Basically, you want to only run checkstyle manually and infrequently, but guard doesn't provide configs for events.

No what I meant is if I configure java linter like this ``require("guard.filetype")("java"):lint("checkstyle")`` It wont run and throw error saying no formatters found.

**Comment by xiaoshihou514** [comment]
This doesn't sound right, I will take a look a bit later.

**Comment by 231tr0n** [comment]
> This doesn't sound right, I will take a look a bit later.

Sure

**Comment by xiaoshihou514** [comment]
![Screenshot_2024-12-09-17-47-36-147_com.termux-edit.jpg](https://github.com/user-attachments/assets/45ea1c75-b762-4d36-83df-a4ed23452482)

Can't seem to repro (linter works standalone as in screenshot)

**Comment by 231tr0n** [comment]
> ![Screenshot_2024-12-09-17-47-36-147_com.termux-edit.jpg](https://private-user-images.githubusercontent.com/108414369/393945652-45ea1c75-b762-4d36-83df-a4ed23452482.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM3NjY4MjcsIm5iZiI6MTczMzc2NjUyNywicGF0aCI6Ii8xMDg0MTQzNjkvMzkzOTQ1NjUyLTQ1ZWExYzc1LWI3NjItNGQzNi04M2RmLWE0ZWQyMzQ1MjQ4Mi5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwOVQxNzQ4NDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04YWY1ODFjZGU1NjI3Zjc0ZmRjZjZlNGQyN2JkZDEyNTMwMjVhNTRjNTJlYjMwOWJhOTlkMDg0OGRhYzU4ZjBiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.BbclIa2DZOChSmPV1SJunp-KgM4U5uhLVMNW7kgIi8c)
> 
> Can't seem to repro

Nvm I found out what the issue was! I did not push checkstyle changes to my forked repo while developing linter for checkstyle! Really sorry for making you debug this issue.

**Comment by 231tr0n** [comment]
> > ![Screenshot_2024-12-09-17-47-36-147_com.termux-edit.jpg](https://private-user-images.githubusercontent.com/108414369/393945652-45ea1c75-b762-4d36-83df-a4ed23452482.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM3NjY4MjcsIm5iZiI6MTczMzc2NjUyNywicGF0aCI6Ii8xMDg0MTQzNjkvMzkzOTQ1NjUyLTQ1ZWExYzc1LWI3NjItNGQzNi04M2RmLWE0ZWQyMzQ1MjQ4Mi5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwOVQxNzQ4NDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04YWY1ODFjZGU1NjI3Zjc0ZmRjZjZlNGQyN2JkZDEyNTMwMjVhNTRjNTJlYjMwOWJhOTlkMDg0OGRhYzU4ZjBiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.BbclIa2DZOChSmPV1SJunp-KgM4U5uhLVMNW7kgIi8c)
> > Can't seem to repro
> 
> Nvm I found out what the issue was! I did not push checkstyle changes to my forked repo while developing linter for checkstyle! Really sorry for making you debug this issue.

![image](https://github.com/user-attachments/assets/9a2b12cf-9366-40f9-9625-7080ff00a90a)
Ok! I am able to reproduce it. Just run ``Guard fmt`` with the following config
![image](https://github.com/user-attachments/assets/5f85578b-77d4-4aeb-a38c-ceacdefab4af)

**Comment by xiaoshihou514** [comment]
Well... You didn't specify a formatter, so obviously `Guard fmt` can't do anything useful...

**Comment by 231tr0n** [comment]
> Well... You didn't specify a formatter, so obviously `Guard fmt` can't do anything useful...

Yeah But I configured a linter right. So how do I get it to lint is my doubt.

**Comment by 231tr0n** [comment]
So there is no way to run a linter without configuring a formatter? Is it possible to provide a ``Guard lint`` to only lint and make ``Guard fmt`` to only format. Probably just invoking ``Guard`` can do both linting and formatting. I think this is a really good idea.

**Comment by xiaoshihou514** [comment]
Hey the linter is triggered automatically on these events

https://github.com/nvimdev/guard.nvim/blob/d5b4a492ed77107eef43261c2b76124583c27aea/lua/guard/filetype.lua#L90-L94

![image](https://github.com/user-attachments/assets/689d5d00-e3cf-4892-a7e8-6d39ce0fdd9e)

Maybe inject a few print statements into the linter to see if your regex is working properly.

You can trigger a manual lint refresh by `:lua require("guard.lint").do_lint()`

Hope that helps :)

**Reply by 231tr0n** [reply]
> Hey the linter is triggered automatically on these events
> 
> https://github.com/nvimdev/guard.nvim/blob/d5b4a492ed77107eef43261c2b76124583c27aea/lua/guard/filetype.lua#L90-L94

The thing is I would like to disable automatic linting. I only enable automatic formatting. Is it really out of scope to provide something as suggested here.

Make ``Guard fmt`` run all configured formatters.
Make ``Guard lint`` run all configured linters.
Make ``Guard`` run all formatters first and linters next.

I think this could be a fairly clean API to do these.

**Reply by xiaoshihou514** [reply]
> Basically, you want to only run checkstyle manually and infrequently, but guard doesn't provide configs for events.

Ah, so it _was_ what I was thinking.

It sounds reasonable, it's just no one has been using a "heavy" linter like checkstyle, please give me a few days to think about this.

**Reply by 231tr0n** [reply]
Also can you please consider all the other requests I made as well. I feel these are fairly important functionality to add. Mainly the synchronous formatting ones.
2. Provide Guard lint to only lint and not format file. Also make Guard fmt to only format and not lint file.
3. Can I have a combination where linter is run asynchronously and formatter is run synchronously.
4. Provide options for configuring async nature of linters and formatters in vim.g.guard_config.

**Reply by xiaoshihou514** [reply]
> There is no way to run the linter if a formatter is not configured.

There is, just not a formally defined command interface. `:lua require("guard.lint").do_lint()`

> Provide Guard lint to only lint and not format file. Also make Guard fmt to only format and not lint file.

Same as above, I am hesitant to make `Guard lint` public API because I don't think many would benefit from this.

> Can I have a combination where linter is run asynchronously and formatter is run synchronously.
> Provide options for configuring async nature of linters and formatters in vim.g.guard_config.

Making guard sync is a lot of work and it's unclear how one exposes the interface. Also I don't think many would prefer freezing the entire editor when doing sync-ed formatting.

Perhaps you can use the guard user autocmds to lock the buffer as a workaround?

**Reply by 231tr0n** [reply]
Well ``Guard lint`` is something which many people would benifit afaik if they work on large code bases with a single file consisting of lines more than 40000k and plus and with heavy linters like checkstyle. Like it just takes a lot of time to lint big codebases.

I also wanted sync-ed formatting since for large codebases formatting make take some time probably 3 - 4 seconds and if we make any changes in the mean time in buffer, all of them are gone.

This way my intention behind asking for commands like these
Make ``Guard fmt`` run all configured formatters.
Make ``Guard lint`` run all configured linters.
Make ``Guard`` run all formatters first and linters next.

Anyways the thing is without these features, I dont think I can actually make good use of the functionality of this plugin at work. That was my pain point behind raising these concerns. 
I apologize if I am being too persistent on these changes.

**Reply by xiaoshihou514** [reply]
I decided to add `Guard lint` to make the API more symmetric, probably `disable-fmt` and `disable-lint` as well. The final solution here would be providing an API for users to tinker with autocmds. (You might need to `nvim_del_autocmd` as a workaround for now).

But I insist no sync-ed formatting features would be implemented.
- Changing the buffer during formatting is clearly a race condition which leads to undefined behaviour.
- You can use autocmds to lock the buffer (`set modifiable`) instead of freezing the whole editor
- If you don't plan to do work during formatting, you might as well do it in the cmdline
- Sync-ed formatting is easy and one could easily use `:!` to implement their own, this is not the purpose of guard

**Reply by 231tr0n** [reply]
> I decided to add `Guard lint` to make the API more symmetric, probably `disable-fmt` and `disable-lint` as well. The final solution here would be providing an API for users to tinker with autocmds. (You might need to `nvim_del_autocmd` as a workaround for now).
> 
> But I insist no sync-ed formatting features would be implemented.
> - Changing the buffer during formatting is clearly a race condition which leads to undefined behaviour.
> - You can use autocmds to lock the buffer (`set modifiable`) instead of freezing the whole editor
> - If you don't plan to do work during formatting, you might as well do it in the cmdline
> - Sync-ed formatting is easy and one could easily use `:!` to implement their own, this is not the purpose of guard

Make ``Guard fmt`` run all configured formatters.
Make ``Guard lint`` run all configured linters.
Make ``Guard`` run all formatters first and linters next
Is this how the new api is going to work?

I am asking this because I dont want linter to be run when running ``Guard fmt``.

**Reply by xiaoshihou514** [reply]
Just `Guard lint`, use `Guard fmt | Guard lint` if you want both.

`Guard fmt` doesn't run linters, it's just currently configured linters listen to the guard user event. (again, you can delete these autocmds if you want)

**Reply by xiaoshihou514** [reply, marked as answer]
Done as of `2.0.0`

**Reply by 231tr0n** [reply]
Awesome, I will integrate support for checkstyle editor tomorrow and raise PR for it. Thank you for considering all the changes.
